File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/camera/RCTMGLCameraManager.java
package com.mapbox.rctmgl.components.camera;

import com.mapbox.geojson.FeatureCollection;
import com.facebook.common.logging.FLog;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.mapbox.rctmgl.components.AbstractEventEmitter;
import com.mapbox.rctmgl.utils.GeoJSONUtils;
//import com.mapbox.rctmgl.utils.GeoJSONUtils;

import java.util.HashMap;
import java.util.Map;

public class RCTMGLCameraManager extends AbstractEventEmitter<RCTMGLCamera> {
    public static final String REACT_CLASS = "RCTMGLCamera";

    private ReactApplicationContext mContext;

    public RCTMGLCameraManager(ReactApplicationContext context) {
        super(context);
        mContext = context;
    }

    @Override
    public Map<String, String> customEvents() {
        return new HashMap<>();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected RCTMGLCamera createViewInstance(ThemedReactContext reactContext) {
        return new RCTMGLCamera(reactContext, this);
    }

    @ReactProp(name="stop")
    public void setStop(RCTMGLCamera camera, ReadableMap map) {
        if (map != null) {
            CameraStop stop = CameraStop.fromReadableMap(mContext, map, null);
            camera.setStop(stop);
        }
    }

    @ReactProp(name="defaultStop")
    public void setDefaultStop(RCTMGLCamera camera, ReadableMap map) {
        if (map != null) {
            CameraStop stop = CameraStop.fromReadableMap(mContext, map, null);
            camera.setDefaultStop(stop);
        }
    }
/*v10todo
    @ReactProp(name="maxBounds")
    public void setMaxBounds(RCTMGLCamera camera, String value) {
        if (value != null) {
            FeatureCollection collection = FeatureCollection.fromJson(value);
            camera.setMaxBounds(GeoJSONUtils.toLatLngBounds(collection));
        }
    }
*/

    @ReactProp(name="userTrackingMode")
    public void setUserTrackingMode(RCTMGLCamera camera, int userTrackingMode) {
        camera.setUserTrackingMode(userTrackingMode);
        throw new AssertionError("Unused code");
    }

    @ReactProp(name="followZoomLevel")
    public void setZoomLevel(RCTMGLCamera camera, double zoomLevel) {
        camera.setZoomLevel(zoomLevel);
    }

    @ReactProp(name="followUserLocation")
    public void setFollowUserLocation(RCTMGLCamera camera, boolean value) {
        camera.setFollowUserLocation(value);
    }

    @ReactProp(name="followUserMode")
    public void setFollowUserMode(RCTMGLCamera camera, String value) {
        camera.setFollowUserMode(value);
    }

    @ReactProp(name="minZoomLevel")
    public void setMinZoomLevel(RCTMGLCamera camera, double value) {
        camera.setMinZoomLevel(value);
    }

    @ReactProp(name="maxZoomLevel")
    public void setMaxZoomLevel(RCTMGLCamera camera, double value) {
        camera.setMaxZoomLevel(value);
    }

    @ReactProp(name="followPitch")
    public void setFollowPitch(RCTMGLCamera camera, double value) {
        camera.setFollowPitch(value);
    }

}

File: android/rctmgl/src/main/java-v10/com/mapbox/rctmgl/components/camera/RCTMGLCameraManager.kt
package com.mapbox.rctmgl.components.camera

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.mapbox.geojson.FeatureCollection
import com.mapbox.rctmgl.components.AbstractEventEmitter
import com.mapbox.rctmgl.components.camera.CameraStop.Companion.fromReadableMap
import com.mapbox.rctmgl.utils.GeoJSONUtils.toLatLngBounds


//import com.mapbox.rctmgl.utils.GeoJSONUtils;
class RCTMGLCameraManager(private val mContext: ReactApplicationContext) :
    AbstractEventEmitter<RCTMGLCamera?>(
        mContext
    ) {
    override fun customEvents(): Map<String, String>? {
        return HashMap()
    }

    override fun getName(): String {
        return REACT_CLASS
    }

    override fun createViewInstance(reactContext: ThemedReactContext): RCTMGLCamera {
        return RCTMGLCamera(reactContext, this)
    }

    @ReactProp(name = "stop")
    fun setStop(camera: RCTMGLCamera, map: ReadableMap?) {
        if (map != null) {
            val stop = fromReadableMap(mContext, map, null)
            camera.setStop(stop)
        }
    }

    @ReactProp(name = "defaultStop")
    fun setDefaultStop(camera: RCTMGLCamera, map: ReadableMap?) {
        if (map != null) {
            val stop = fromReadableMap(mContext, map, null)
            camera.setDefaultStop(stop)
        }
    }

    @ReactProp(name = "userTrackingMode")
    fun setUserTrackingMode(camera: RCTMGLCamera, userTrackingMode: Int) {
        camera.setUserTrackingMode(userTrackingMode)
        throw AssertionError("Unused code")
    }

    @ReactProp(name = "zoomLevel")
    fun setZoomLevel(camera: RCTMGLCamera, zoomLevel: Double) {
        camera.setZoomLevel(zoomLevel)
    }

    @ReactProp(name = "followUserLocation")
    fun setFollowUserLocation(camera: RCTMGLCamera, value: Boolean) {
        camera.setFollowUserLocation(value)
    }

    @ReactProp(name = "followUserMode")
    fun setFollowUserMode(camera: RCTMGLCamera, value: String?) {
        camera.setFollowUserMode(value)
    }

    @ReactProp(name = "minZoomLevel")
    fun setMinZoomLevel(camera: RCTMGLCamera, value: Double) {
        camera.setMinZoomLevel(value)
    }

    @ReactProp(name = "maxZoomLevel")
    fun setMaxZoomLevel(camera: RCTMGLCamera, value: Double) {
        camera.setMaxZoomLevel(value)
    }

    @ReactProp(name = "followPitch")
    fun setFollowPitch(camera: RCTMGLCamera, value: Double) {
        camera.setFollowPitch(value)
    }

    @ReactProp(name = "followHeading")
    fun setFollowHeading(camera: RCTMGLCamera, value: Double) {
        camera.setFollowHeading(value)
    }

    @ReactProp(name = "followZoomLevel")
    fun setFollowZoomLevel(camera: RCTMGLCamera, value: Double) {
        camera.setFollowZoomLevel(value)
    }

    @ReactProp(name = "maxBounds")
    fun setMaxBounds(camera: RCTMGLCamera, value: String?) {
        if (value != null) {
            val collection = FeatureCollection.fromJson(value)
            camera.setMaxBounds(toLatLngBounds(collection))
        } else {
            camera.setMaxBounds(null)
        }
    }

    companion object {
        const val REACT_CLASS = "RCTMGLCamera"
    }
}